### ISSUE TITLE
Preserve whitespace when braces are omitted

### ISSUE THREAD

**Issue by thysultan** [prettier/prettier#3049]
**Prettier 1.7.4**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22function%20forEach%20(callback)%20%7B%5Cn%5Ctfor%20(var%20i%20%3D%200%3B%20i%20%3C%2010%3B%20%2B%2Bi)%5Cn%5Ct%5Ctif%20(true)%5Cn%5Ct%5Ct%5Ctcallback(i)%5Cn%7D%5Cn%5Cnfunction%20assign(object%2C%20primary%2C%20secondary)%20%7B%5Cn%5Ctfor%20(var%20key%20in%20primary)%5Cn%5Ct%5Ctobject%5Bkey%5D%20%3D%20primary%5Bkey%5D%5Cn%5Cn%5Ctfor%20(var%20key%20in%20secondary)%5Cn%5Ct%5Ctobject%5Bkey%5D%20%3D%20secondary%5Bkey%5D%5Cn%5Ct%5Cn%5Ctreturn%20object%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A100%2C%22semi%22%3Afalse%2C%22singleQuote%22%3Atrue%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--print-width 100
--no-semi
--single-quote
```

**Input:**
```jsx
function forEach (callback) {
	for (var i = 0; i < 10; ++i)
		if (true)
			callback(i)
}

function assign(object, primary, secondary) {
	for (var key in primary)
		object[key] = primary[key]

	for (var key in secondary)
		object[key] = secondary[key]
	
	return object
}
```

**Output:**
```jsx
function forEach(callback) {
  for (var i = 0; i < 10; ++i) if (true) callback(i)
}

function assign(object, primary, secondary) {
  for (var key in primary) object[key] = primary[key]

  for (var key in secondary) object[key] = secondary[key]

  return object
}

```

**Expected behavior:**

```js
function forEach(callback) {
  for (var i = 0; i < 10; ++i) 
  	if (true)
  		callback(i)
}

function assign(object, primary, secondary) {
  for (var key in primary)
  	object[key] = primary[key]

  for (var key in secondary)
  	object[key] = secondary[key]

  return object
}
```

**Comment by kasiriveni** [comment]
+1

**Comment by suchipi** [comment]
I don't omit braces much, but when I do, it's generally only for "guard conditions":
```js
function foo() {
  if (bar) return;

  // ...
}
```

In that case, I like that the expression is only one line. But this looks weird:
```js
for (var i = 0; i < 10; ++i) if (true) callback(i)
```

Is it worth treating those two things differently? That is... we could make it indent when braces are omitted for non-returns, but if it's a return, then we inline it

**Comment by thysultan** [comment]
@suchipi I think it should behave like it would without the braces. [Playground link].(https://prettier.io/playground/#%7B%22content%22%3A%22function%20foo()%20%7B%5Cn%20%20if%20(bar)%20return%5Cn%20%20if%20(bar)%5Cn%20%20%20%20return%5Cn%20%20%5Cn%20%20%2F%2F%20should%20prettify%20like%20it%20would%20this%2C%20minus%20braces%5Cn%20%20if%20(bar)%20%7Breturn%7D%5Cn%20%20if%20(bar)%20%7B%5Cn%20%20%20%20return%5Cn%20%20%7D%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Afalse%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)

**Input**

```js
function foo() {
  if (bar) return
  if (bar)
    return
  
  // should prettify like it would this, minus braces
  if (bar) {return}
  if (bar) {
    return
  }
}
```

**Output**

```js
function foo() {
  if (bar) 
    return
  if (bar) 
    return

  // should prettify like it would this, minus braces
  if (bar) {
    return
  }
  if (bar) {
    return
  }
}

```

**Comment by azz** [comment]
A lot of people strongly prefer

```js
if (bar) return
```

to

```js
if (bar)
  return
```

so that would be a highly controversial change.

**Comment by suchipi** [comment]
Considering @azz's input, and that `if (bar) return` is what prettier currently prints, I think we shouldn't change the way that particular pattern is printed.

I'm open to changing the behavior for for loops and *maybe* if statements whose body is not a return statement... but maybe if the way prettier inlines it looks weird to you, it's an indication that you might want to use braces? 🙊

**Comment by thysultan** [comment]
My main gripe with the `if (condition) return` output is the lack of consistency between braces and no-braces – that is you might expect it to apply the same treatment to both patterns – `if (condition) ..inline..` and `if (condition) {..inline..}` yielding the same opinionated format:

```js
if (condition) {
  ..nested..
}
```

Be that as it may, the original issue is more about the uncommon case of outputting patterns that might unintentionally decrease the readability of the original code.

**Comment by mnpenner** [comment]
I personally like it on one line. If it's on the next line and indented someone might try adding a 2nd statement to the "block" not realizing there are no braces.

If you want it on the next line, I think you should have to add braces, and formatting it this way encourages that.

**Comment by The-Smallest** [comment]
> If it's on the next line and indented someone might try adding a 2nd statement to the "block" not realizing there are no braces.

@mnpenner Well, we have *prettier* to solve this kind of problems. After reformatting it would be out of "block".

> As this is controversial we could add an option like --conditions-no-one-liner.

This would be really nice.

90% of my code has only one statement in _for_ or _if_.


```javascript
function canRead(projectId) {
  if (isAdmin)
    return true

  if (myProjects[projectId])
    return false

  if (someCheckFunction(projectId) || anotherLinkedCheck())
    return true

  return false
}
```

I could rewrite it using `if (cond) statement`, but I've intentionally split this rules: it is easy to scan this code with your eyes.

*condition*
*result* - 1 line below

Simple _top-to-bottom_ scan, meanwhile _singleline if_ forces to use *Z* eye-scan-pattern.
Compare:
```javascript
function canRead(projectId) {
  if (isAdmin) return true

  if (myProjects[projectId]) return false

  if (someCheckFunction(projectId) || anotherLinkedCheck()) return true

  return false
}
```

I was forced to use prettier (_Vetur_ for VSCode has switched to it). 
And now I have to insert a lot of new braces (which I don't really like) *manually* to avoid _if-inlining_.
Or dirty-hack it with comment.

**Comment by mnpenner** [comment]
@The-Smallest True... but you might not notice the re-alignment unless you're diligently looking over your commits.

I do see what you mean about the return being on the next line is a little easier to read in that scenario.

I guess a flag would be best here.
